### PR TITLE
Inbound 99DOTS integration

### PR DESCRIPTION
**Pull request by proteusvacuum**
This allows 99DOTS to make calls to Commcare to create and update adherence cases
https://docs.google.com/document/d/1PH4EPcSqhsYP-lV-_4_V6C9AvkpaBmIEe5ID5U2mAlc/edit#heading=h.d4o2i1xsr1tf

@czue @kaapstorm

### CODE REVIEW

File: custom/enikshay/integrations/ninetyninedots/tests/test_integration.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        self.assertItemsEqual(
+            [parse_datetime(adherence_time) for adherence_time in adherence_times],
+            [parse_datetime(adherence_value['timestamp']) for adherence_value in adherence_values]
+        )
+
+        for adherence_case in adherence_cases:
+            self.assertEqual(
+                adherence_case.dynamic_case_properties().get('adherence_confidence'),
+                'high'
+             )
```
Comment (dimagimon) [inline custom/enikshay/integrations/ninetyninedots/tests/test_integration.py]: E121 continuation line under-indented for hanging indent

**Comment by czue** [comment]
build is failing with various instances of
```
======================================================================
FAIL: test_force_consumption (corehq.apps.commtrack.tests.test_xml.CommTrackOTATest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 232, in inner
    return helper(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 205, in __call__
    call_with_settings(fn_with_pre_and_post, run_config.settings, args, kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 220, in call_with_settings
    fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 201, in fn_with_pre_and_post
    self.fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/commtrack/tests/test_xml.py", line 161, in test_force_consumption
    balance_blocks = _get_ota_balance_blocks(self.domain, self.user)
  File "/mnt/commcare-hq-ro/corehq/apps/commtrack/tests/test_xml.py", line 708, in _get_ota_balance_blocks
    return extract_balance_xml(restore_config.get_payload().as_string())
  File "corehq/ex-submodules/casexml/apps/phone/restore.py", line 601, in get_payload
    response = self.generate_payload()
  File "corehq/ex-submodules/casexml/apps/phone/restore.py", line 607, in generate_payload
    response = self._generate_restore_response(async_task=async_task)
  File "corehq/ex-submodules/casexml/apps/phone/restore.py", line 673, in _generate_restore_response
    partial_response = provider.get_response(self.restore_state)
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/providers.py", line 12, in get_response
    return sync_op.get_payload()
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py", line 258, in get_payload
    case_ids_to_sync = self.get_case_ids_to_sync()
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py", line 265, in get_case_ids_to_sync
    case_ids_to_sync = case_ids_to_sync | set(self.get_case_ids_for_owner(owner_id))
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py", line 284, in get_case_ids_for_owner
    return self._get_case_ids_for_owners_without_extensions(owner_id)
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py", line 300, in _get_case_ids_for_owners_without_extensions
    return get_case_footprint_info(self.restore_state.domain, owner_id).all_ids
  File "corehq/ex-submodules/casexml/apps/phone/cleanliness.py", line 240, in get_case_footprint_info
    dependent_cases = get_dependent_case_info(domain, open_case_ids)
  File "corehq/ex-submodules/casexml/apps/phone/cleanliness.py", line 208, in get_dependent_case_info
    direct_dependencies = _get_direct_dependencies(domain, cases)
  File "corehq/ex-submodules/casexml/apps/phone/cleanliness.py", line 221, in _get_direct_dependencies
    extension_cases = set(case_accessor.get_extension_case_ids(cases))
  File "/mnt/commcare-hq-ro/corehq/form_processor/interfaces/dbaccessors.py", line 314, in get_extension_case_ids
    return self.db_accessor.get_extension_case_ids(self.domain, case_ids)
  File "/mnt/commcare-hq-ro/corehq/form_processor/backends/couch/dbaccessors.py", line 164, in get_extension_case_ids
    return get_extension_case_ids(domain, case_ids)
  File "corehq/ex-submodules/casexml/apps/case/dbaccessors/related.py", line 24, in get_extension_case_ids
    return [r.case_id for r in get_all_reverse_indices_info(domain, case_ids, CASE_INDEX_EXTENSION)]
  File "corehq/ex-submodules/casexml/apps/case/dbaccessors/related.py", line 61, in get_all_reverse_indices_info
    keys=_get_keys_for_reverse_index_view(domain, case_ids, relationship),
  File "corehq/ex-submodules/casexml/apps/case/dbaccessors/related.py", line 67, in _get_keys_for_reverse_index_view
    assert isinstance(case_ids, list)
AssertionError: 
```

File: custom/enikshay/integrations/ninetyninedots/utils.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            'default_adherence_confidence', self.DEFAULT_ADHERENCE_CONFIDENCE
+        )
+
+    @property
+    @memoized
+    def _episode_case(self):
+        return get_open_episode_case(self.domain, self._person_case.case_id)
+
+    def create_adherence_cases(self, adherence_points, adherence_source):
+        return self.case_factory.create_or_update_cases([
```
Comment (czue) [inline custom/enikshay/integrations/ninetyninedots/utils.py]: was there a reason to use `CaseFactory` over `CaseBlock`? (not saying it matters, more just curious)

**Comment by czue** [comment]
i didn't review very closely since it's pretty custom stuff but looks good. nice tests.  :+1: on :white_check_mark: from my side

are you working on the forwarding stuff separately?

File: custom/enikshay/integrations/ninetyninedots/utils.py (hunk 2 above)
Comment (proteusvacuum) [inline custom/enikshay/integrations/ninetyninedots/utils.py]: No reason really, just more used to the syntax at this point.

**Comment by proteusvacuum** [comment]
Yeah, forwarding stuff forthcoming - wanted to get this in first so I could send it to Nakull to test.

**Comment by kaapstorm** [comment]
LGTM. (Also enjoyed the Hobbitization of the tests. :volcano: )

**Comment by proteusvacuum** [comment]
✅

**Pull request merged**